karoo_gp/base_class.py

```diff
@@ -1787,1535 +2043,1588 @@ class labels provided in the data .csv. Outputs an array of tuples
+    Called by: fx_evolve_child_link_fix, fx_evolve_banch_top_copy,
+                fx_evolve_branch_body_copy
 
-            elif int(tree[8][node]) == 3:  # if arity = 3
-                tree[9][node] = c_buffer
-                tree[10][node] = c_buffer + 1
-                tree[11][node] = c_buffer + 2
+    Arguments required: tree, node
+    '''
+    parent_arity_sum = 0
+    prior_sibling_arity = 0
+    prior_siblings = 0
 
-            else:
-                print('\n\t\033[31m ERROR! In fx_evolve_child_link: node',
-                      node, 'has arity', tree[8][node])
-                # consider special instructions for this (pause) - 2019 06/08
-                self.fx_karoo_pause()
+    # increment through all nodes (exclude 0) in array 'tree'
+    for n in range(1, len(tree.root[3])):
 
-        return tree
+        # find parent nodes at the prior depth
+        if int(tree.root[4][n]) == int(tree.root[4][node])-1:
+            if tree.root[8][n] != '':
+                # sum arities of all parent nodes at the prior depth
+                parent_arity_sum = parent_arity_sum + int(tree.root[8][n])
 
+        # find prior siblings at the current depth
+        if (int(tree.root[4][n]) == int(tree.root[4][node]) and
+            int(tree.root[3][n]) < int(tree.root[3][node])):
+            if tree.root[8][n] != '':
+                # sum prior sibling arity
+                prior_sibling_arity = prior_sibling_arity + int(tree.root[8][n])
+            # sum quantity of prior siblings
+            prior_siblings = prior_siblings + 1
 
-    def fx_evolve_child_link_fix(self, tree):
+    # One algo to rule the world!
+    c_buffer = node + (parent_arity_sum + prior_sibling_arity - prior_siblings)
 
-        '''
-        In a given Tree, fix 'node_c1', 'node_c2', 'node_c3' for all nodes.
+    return c_buffer
 
-        This is required anytime the size of the array 'gp.tree' has been
-        modified, as with both Grow and Full mutation.
+# used by: Population
+def fx_evolve_child_link(tree, node, c_buffer, error):
 
-        Called by: fx_evolve_grow_mutate, fx_evolve_crossover,
-                   fx_evolve_branch_body_copy, fx_evolve_branch_copy
+    '''
+    Link each parent node to its children.
 
-        Arguments required: tree
-        '''
+    Called by: fx_evolve_child_link_fix
 
-        # tested 2015 06/04
-        for node in range(1, len(tree[3])):
+    Arguments required: tree, node, c_buffer
+    '''
 
-            # generate c_buffer for each node
-            c_buffer = self.fx_evolve_c_buffer(tree, node)
-            # update child links for each node
-            tree = self.fx_evolve_child_link(tree, node, c_buffer)
+    if int(tree.root[3][node]) == 1:
+        # if root (node 1) is passed through this method
+        c_buffer = c_buffer + 1
 
-        return tree
+    if tree.root[8][node] != '':
 
+        if int(tree.root[8][node]) == 0:  # if arity = 0
+            tree.root[9][node] = ''
```

Comment:
Ah, that's probably why.  As I mentioned at the beginning, I don't like too much to have a field that can be a `str` or an `int`.